src/store/compression_lz4_block.rs

```diff
@@ -0,0 +1,39 @@
+use std::io::{self};
+
+use core::convert::TryInto;
+use lz4_flex::{compress_into, decompress_into};
+/// Name of the compression scheme used in the doc store.
+///
+/// This name is appended to the version string of tantivy.
+pub const COMPRESSION: &str = "lz4_block";
+
+pub fn compress(uncompressed: &[u8], compressed: &mut Vec<u8>) -> io::Result<()> {
+    compressed.clear();
+
+    compressed.extend_from_slice(&[0, 0, 0, 0]);
+    compress_into(uncompressed, compressed);
+    let size = uncompressed.len() as u32;
+    compressed[0] = size as u8;
+    compressed[1] = (size >> 8) as u8;
+    compressed[2] = (size >> 16) as u8;
+    compressed[3] = (size >> 24) as u8;
+    Ok(())
+}
+
+pub fn decompress(compressed: &[u8], decompressed: &mut Vec<u8>) -> io::Result<()> {
+    decompressed.clear();
+    //next lz4_flex version will support slice as input parameter.
+    //this will make the usage much less ugly
+    let size = compressed.get(..4).ok_or(io::ErrorKind::InvalidData)?;
+    let size: &[u8; 4] = size.try_into().unwrap();
+    let uncompressed_size = u32::from_le_bytes(*size) as usize;
+    // reserve more than required, because blocked writes may write out of bounds, will be improved
+    // with lz4_flex 1.0
+    decompressed.reserve(uncompressed_size + 4 + 24);
+    unsafe {
+        decompressed.set_len(uncompressed_size);
+    }
+    decompress_into(&compressed[4..], decompressed)
```

Comment:
I am extremely confused here.
It seems like decompress_into is expecting a `&mut Vec`.
Aren't we supposed to pass it a clear one? What is the point of set_len?

Reply:
Ok I look into lz4_flex to understand... I think I understood the gist of it.

The API has the same name but is very different between safe and unsafe, which got me confused.

Reply:
sorry for that, I just applied my knowledge here to avoid a copy. the next lz4_flex version with support for slices should improve this

Reply:
slice API will require you do special case the last part?

Or will the the API will look like :
- pass a slice that has decompressed_len + extra len and trim it done to decompressed_len on your own
or
- pass a slice that has decompressed_len but it needs to be from a vec with an extra capacity.

An alternative could be 
decompress(compress: &[u8], decompressed: &mut Vec<u8>, len_upperbound: usize);

Reply:
The slice api will accept a slice in the correct length for uncompressed data. The blocked writes will be disabled for the last bytes, so that there will be no out of bounds writes. It's also an requirement for the frame API or blocks next to each other could accidentally overwrite data 
|Block 1| Block 2 | ..

Reply:
Ok!